zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/behavior/BpmnJobBehavior.java

```diff
@@ -159,9 +166,98 @@ private Either<Failure, List<LinkedResourceProps>> evalLinkedResourceProps(
+    };
+  }
+
+  private Either<Failure, PersistedResource> findResourceByIdInSameDeployment(
+      final String resourceId, final BpmnElementContext context, final long scopeKey) {
+    return stateBehavior
+        .getDeploymentKey(context.getProcessDefinitionKey(), context.getTenantId())
+        .flatMap(
+            deploymentKey ->
+                resourceState
+                    .findResourceByIdAndDeploymentKey(
+                        resourceId, deploymentKey, context.getTenantId())
+                    .<Either<Failure, PersistedResource>>map(Either::right)
+                    .orElseGet(
+                        () ->
+                            Either.left(
+                                new Failure(
+                                    String.format(
+                                        """
+                                        Expected to use a resource with id '%s' with binding type 'deployment', \
+                                        but no such resource found in the deployment with key %s which contained the current process. \
+                                        To resolve this incident, migrate the process instance to a process definition \
+                                        that is deployed together with the intended resource to use.\
+                                        """,
+                                        resourceId, deploymentKey),
+                                    ErrorType.RESOURCE_NOT_FOUND,
+                                    scopeKey))));
+  }
+
+  private Either<Failure, PersistedResource> findLatestResourceById(
+      final String resourceId, final String tenantId, final long scopeKey) {
+    return resourceState
+        .findLatestResourceById(resourceId, tenantId)
+        .<Either<Failure, PersistedResource>>map(Either::right)
+        .orElseGet(
+            () ->
+                Either.left(
+                    new Failure(
+                        String.format(
+                            """
+                            Expected to find a resource with id '%s', but no resource with this id is found, \
+                            at least a resource with this id should be available. \
+                            To resolve the Incident please deploy a resource with the same id.
+                            """,
+                            resourceId),
+                        ErrorType.RESOURCE_NOT_FOUND,
+                        scopeKey)));
+  }
+
+  private Either<Failure, PersistedResource> findResourceByIdAndVersionTag(
+      final String resourceId,
+      final String versionTag,
+      final String tenantId,
+      final long scopeKey) {
+    return resourceState
+        .findResourceByIdAndVersionTag(resourceId, versionTag, tenantId)
+        .<Either<Failure, PersistedResource>>map(Either::right)
+        .orElseGet(
+            () ->
+                Either.left(
+                    new Failure(
```

Comment:
🔧 We have a nice helper method to construct an Either from an Optional which allows you to avoid the juggling with generics: `Either.ofOptional`

```suggestion
    return Either.ofOptional(
            resourceState.findResourceByIdAndVersionTag(resourceId, versionTag, tenantId))
        .orElse(
            new Failure(
```